builtins/web/fetch/headers.h

```diff
@@ -53,34 +54,61 @@ class Headers final : public BuiltinImpl<Headers> {
   /// If a header is added, deleted, or replaced on an instance in `CachedInContent` mode, the
   /// instance transitions to `ContentOnly` mode, and the underlying resource handle is discarded.
   enum class Mode {
-    HostOnly, // Headers are stored in the host.
+    HostOnly,        // Headers are stored in the host.
     CachedInContent, // Host holds canonical headers, content a cached copy.
-    ContentOnly, // Headers are stored in a Map held by the `Entries` slot.
-    Uninitialized, // Headers have not been initialized.
+    ContentOnly,     // Headers are stored in a Map held by the `Entries` slot.
+    Uninitialized,   // Headers have not been initialized.
   };
 
+  // Headers internal data structure is a list of key-value pairs, ready to go on the wire as
+  // owned host strings.
+  using HeadersList = std::vector<std::tuple<host_api::HostString, host_api::HostString>>;
+  // A sort list is maintained of ordered indicies of the the sorted lowercase keys of main headers
+  // list, with each index of HeadersList always being present in this list once and only once.
+  // All lookups are done as indices in this list, which then map to indices in HeadersList.
+  // When this list is empty, that means the sort list is not valid and needs to be computed. For
+  // example, it is cleared after an insertion. It is recomputed lazily for every lookup.
+  using HeadersSortList = std::vector<size_t>;
+
   enum class Slots {
     Handle,
-    Entries, // Map holding headers if they are available in-content.
+    HeadersList,
+    HeadersSortList,
     Mode,
     Guard,
     Count,
   };
 
-  /**
-   * Adds the given header name/value to `self`'s list of headers iff `self`
-   * doesn't already contain a header with that name.
-   */
-  static bool set_if_undefined(JSContext *cx, JS::HandleObject self, string_view name,
-                               string_view value);
+  enum class HeadersGuard {
+    None,
+    Request,
+    Response,
+    Immutable,
+  };
+
+  /// Adds the valid given header name/value to `self`'s list of headers iff `self`
```

Comment:
```suggestion
  /// Adds the valid given header name/value to `self`'s list of headers if `self`
```

Reply:
The "iff" here stands for "[if, and only if](https://en.wikipedia.org/wiki/If_and_only_if)", so it's not a spelling mistake :)